lib/client/doc.js

```diff
@@ -413,30 +410,66 @@ Doc.prototype.fetch = function(callback) {
 
 // Fetch the initial document and keep receiving updates
 Doc.prototype.subscribe = function(callback) {
-  this.wantSubscribe = true;
-  if (this.connection.canSend) {
-    var isDuplicate = this.connection.sendSubscribe(this);
-    pushActionCallback(this.inflightSubscribe, isDuplicate, callback);
-    return;
-  }
-  this.pendingFetch.push(callback);
+  var wantSubscribe = true;
+  this._queueSubscribe(wantSubscribe, callback);
 };
 
 // Unsubscribe. The data will stay around in local memory, but we'll stop
 // receiving updates
 Doc.prototype.unsubscribe = function(callback) {
-  this.wantSubscribe = false;
-  // The subscribed state should be conservative in indicating when we are
-  // subscribed on the server. We'll actually be unsubscribed some time
-  // between sending the message and hearing back, but we cannot know exactly
-  // when. Thus, immediately mark us as not subscribed
-  this.subscribed = false;
+  var wantSubscribe = false;
+  this._queueSubscribe(wantSubscribe, callback);
+};
+
+Doc.prototype._queueSubscribe = function(wantSubscribe, callback) {
+  var nextRequest = this.inflightSubscribe || this.pendingSubscribe[0];
+  var hasOtherRequests = this.pendingSubscribe.some(function(request) {
+    return request !== nextRequest;
+  });
+  if (nextRequest && nextRequest.wantSubscribe === wantSubscribe && !hasOtherRequests) {
+    var callbacks = [];
+    if (nextRequest.callback) callbacks.push(nextRequest.callback);
+    if (callback) callbacks.push(callback);
+    if (callbacks.length) {
+      nextRequest.callback = function(error) {
+        util.callEach(callbacks, error);
+      };
+    }
+    return;
+  }
+  this.pendingSubscribe.push({
+    wantSubscribe: !!wantSubscribe,
+    callback: callback
+  });
+  this._flushSubscribe();
+};
+
+Doc.prototype._flushSubscribe = function() {
+  if (this.inflightSubscribe || !this.pendingSubscribe.length) return;
+
   if (this.connection.canSend) {
-    var isDuplicate = this.connection.sendUnsubscribe(this);
-    pushActionCallback(this.inflightUnsubscribe, isDuplicate, callback);
+    this.inflightSubscribe = this.pendingSubscribe.shift();
+    this.wantSubscribe = this.inflightSubscribe.wantSubscribe;
+    if (this.wantSubscribe) {
+      this.connection.sendSubscribe(this);
+    } else {
+      // Be conservative about our subscription state. We'll be unsubscribed
+      // some time between sending this request, and receiving the callback,
+      // so let's just set ourselves to unsubscribed now.
+      this.subscribed = false;
+      this.connection.sendUnsubscribe(this);
+    }
+
     return;
   }
-  if (callback) process.nextTick(callback);
+
+  if (!this.pendingSubscribe[0].wantSubscribe) {
```

Comment:
Would be nice to add a short comment explaining this case - if the client offline and requests to unsubscribe, immediately call back since offline means it's not subscribed.